src/hotspot/share/runtime/vframeArray.cpp

```diff
@@ -85,19 +85,24 @@ void vframeArrayElement::fill_in(compiledVFrame* vf, bool realloc_failures) {
 
       // Allocate monitor chunk
       _monitors = new MonitorChunk(list->length());
-      vf->thread()->add_monitor_chunk(_monitors);
 
       // Migrate the BasicLocks from the stack to the monitor chunk
       for (index = 0; index < list->length(); index++) {
         MonitorInfo* monitor = list->at(index);
         assert(!monitor->owner_is_scalar_replaced() || realloc_failures, "object should be reallocated already");
         BasicObjectLock* dest = _monitors->at(index);
-        if (monitor->owner_is_scalar_replaced()) {
+        if (monitor->owner_is_scalar_replaced() || monitor->owner() == nullptr) {
```

Comment:
The only way to get a null owner is if owner_is_scalar_replaced() is true:
https://github.com/openjdk/jdk/blob/6bef0474c8b8773d0d20c0f25c36a2ce9cdbd7e8/src/hotspot/share/runtime/stackValue.hpp#L52
and to get this far with it still null means `realloc_failures` is true.  We could avoid a null check later in unpack_on_stack if we skip adding to _monitors in this case.  So maybe use a GrowableArray inside MonitorChunk and add elements using append().
```suggestion
        if (monitor->owner_is_scalar_replaced()) {
```

Reply:
Yes, no nullptr check needed there.
As discussed, moving to a GrowableArray complicates things quite a lot.  fill_in and unpack_on stack both make a similar iteration, but with slightly different information available.  Growing the array on demand and making both loops variable involves more changes in MonitorChunk and its accessors.  I think we're agreed to not be that disruptive for this change.